lib/strings.nix

```diff
@@ -157,6 +157,70 @@ rec {
   */
   replicate = n: s: concatStrings (lib.lists.replicate n s);
 
+  /*
+    Remove leading and trailing whitespace from a string.
+
+    Whitespace is defined as any of the following characters:
+      " ", "\t" "\r" "\n"
+
+    Type: trim :: string -> string
+
+    Example:
+      trim "   hello, world!   "
+      => "hello, world!"
+  */
+  trim = trimWith {
+    start = true;
+    end = true;
+  };
+
+  /*
+    Remove leading and/or trailing whitespace from a string.
+
+    Whitespace is defined as any of the following characters:
+      " ", "\t" "\r" "\n"
+
+    Type: trimWith :: Attrs -> string -> string
+
+    Example:
+      trimWith { start = true; } "   hello, world!   "}
+      => "hello, world!   "
+      trimWith { end = true; } "   hello, world!   "}
+      => "   hello, world!"
+  */
+  trimWith =
+    {
+      # Trim leading whitespace
+      start ? false,
+      # Trim trailing whitespace
+      end ? false,
+      # Forward compatibility
+      ...
```

Comment:
```suggestion
```

I don't think we should have the ellipsis, because if we support an extra argument in the future, the old version shouldn't blindly accept and ignore such an attribute. Instead users should rightly get an error that the old version doesn't accept that attribute.

Reply:
I've seen the pattern used on other functions, but your argument makes sense.

Reply:
Perhaps this one: https://github.com/NixOS/nixpkgs/blob/7af9a82b766258e9cd53e7d1097fe5c748f78672/lib/fileset/default.nix#L799-L804

In that case at least, the difference is that the attributes are provided by the library (and new ones may be added over time), while the user provides the function to handle them. There's something super naughty there to guarantee that the user doesn't forget the `...` in their function: https://github.com/NixOS/nixpkgs/blob/88733d6c42306ab5d094cccdaa8d7f9c90ea63b5/lib/fileset/internal.nix#L835-L837

:laughing:

Reply:
Interesting example!

The only concrete example I can think of is outside of `lib` anyway, so probably not relevant: https://github.com/NixOS/nixpkgs/blob/72b5f58e5bba538ca62c10cd1ab5a4baef03402a/pkgs/applications/editors/neovim/utils.nix#L41-L42

and

https://github.com/NixOS/nixpkgs/blob/b7aed13df5ea800264d72cb76b9db5031afc322f/pkgs/applications/editors/neovim/wrapper.nix#L46

Either way, I've removed the ellipses, thanks for the feedback!